### ISSUE TITLE
Open internal PDF in the modal view

### ISSUE THREAD

**Issue by benoitchantre** [Simbul/baker#609]
We can open an external PDF in the modal view, but not when the file is internal. It works also for a doc file.

When we link the pdf without http, it is loaded inside the current page.

It would be nice to be able to open some local files in the modal view.

Maybe a simple "localhost" in the url could tell to Baker to open the file in the modal view automatically?

**Comment by folletto** [comment]
Just wondering... Doesn't it work with the `file:` protocol?

**Comment by benoitchantre** [comment]
Unfortunately not.

Here's what I have in Xcode output.
The result is the same if the path contains or not the folder name containing the book.

```
2013-01-03 11:20:06.368 myBook[4539:c07]     Page is current page and current page IS NOT delaying loading --> handle clicked link: file://pdf/210x297.pdf?referrer=Baker
2013-01-03 11:20:06.368 myBook[4539:c07]     Page is a link with scheme file:// --> load internal link
2013-01-03 11:20:06.369 myBook[4539:c07] • Page did start load
2013-01-03 11:20:06.371 myBook[4539:c07] • CurrPage failed to load content with error: Error Domain=NSURLErrorDomain Code=-1100 "L’URL demandée est introuvable sur ce serveur." UserInfo=0xa228c40 {NSErrorFailingURLStringKey=file://pdf/210x297.pdf?referrer=Baker, NSErrorFailingURLKey=file://pdf/210x297.pdf?referrer=Baker, NSLocalizedDescription=L’URL demandée est introuvable sur ce serveur., NSUnderlyingError=0x956d560 "L’URL demandée est introuvable sur ce serveur."}
```

**Comment by folletto** [comment]
Ah, yeah. Got it. Ok, thanks for the feedback, we'll look into it. :)

**Comment by benoitchantre** [comment]
If you find a solution to open a local file in the modal view, I think it will be possible also to open other types of files that the iPad can read (word, excel, html, ...)

At the moment, the modal view can display these types of documents but only if they are online.

**Comment by benoitchantre** [comment]
Here's maybe a solution: http://iphonedevsdk.com/forum/iphone-sdk-development/2657-display-a-local-file-in-a-uiwebview.html

**Comment by benoitchantre** [comment]
Here's the technical note from Apple: <a href="http://developer.apple.com/library/ios/ipad/#qa/qa1630/_index.html%23//apple_ref/doc/uid/DTS40008749">Using UIWebView to display select document types</a>.
There's an example how to load a document into a UIWebView.
Does it help?

**Comment by jukialla** [comment]
Hi all. very new to this stuff. Love the framework. I am developing an app with maps. The user can zoom to any scale and [lat.lng]. I would like them to be able to tap on a marker, load a local HTML file as a "popup" with info about the marker (place) they have tapped on and return to the map with the same zoom and lat,lng. Currently, they have to load a new page and return to a new ('refreshed") page that referred them. I can save the zoom and location and restore the referring page, but its a bit clumsy. I would love to be able to load a local HTML page in the model window. I don't want these pages as part of the 'book'. Am I missing something?

**Comment by Simbul** [comment]
@jukialla sounds like something it may be easier to do in a single page with a JS-based solution.

**Comment by jukialla** [comment]
Yes, I have done it with ColorBox in the end. It works fine. Sometimes it throws a low memory warning on iPad 1, but I will continue to fine tune it.

**Comment by Simbul** [comment]
Great. I'm closing this issue then :)

**Comment by benoitchantre** [comment]
@Simbul, this issue is about the ability to open a PDF in the modal view, can you reopen it?

**Comment by Simbul** [comment]
Whoops you're right ;)

**Comment by benoitchantre** [comment]
Thanks.

**Comment by benoitchantre** [comment]
I think that the domain error that happen when I use the _file_ protocol is caused because there's no verification if the path provided is internal or not.

There's a method _isFileReferenceURL_ that returns whether the URL is a file reference URL.
<a href="http://developer.apple.com/library/ios/#documentation/Cocoa/Reference/Foundation/Classes/NSURL_Class/Reference/Reference.html">NSURL Class Reference</a>

Does it help?

**Comment by benoitchantre** [comment]
Here's what happen with the latest code from the master branch
- External PDF with Baker referrer
  `<a href="http://manuals.info.apple.com/fr_FR/ipad_guide_de_l_utilisateur.pdf?referrer=Baker">Guide de l'utilisateur iPad</a>`
  Log: `Opening in Modal Panel with URL: http://manuals.info.apple.com/fr_FR/ipad_guide_de_l_utilisateur.pdf`
  Result: opens the PDF as expected in the modal view
- Local PDF without Baker referrer
  `<a href="pdf/ipad_guide_de_l_utilisateur.pdf">Guide de l'utilisateur iPad (lien interne)</a>`
  Log: `Page is a link with scheme file:// --> load internal link`
  Result: loads the PDF in the page (you can't close the PDF)
- Local PDF with Baker referrer
  `<a href="pdf/ipad_guide_de_l_utilisateur.pdf?referrer=Baker">Guide de l'utilisateur iPad (lien interne)</a>`
  Log: `Page is a link with scheme file:// --> load internal link`
  Result: **loads the PDF in the page instead of opening it in the modal view**

Would it be possible to improve the opening of the modal view when a link has a `referrer` attribute equal to `Baker`?

**Comment by folletto** [comment]
There was still no work done about this, sorry.

**Comment by benoitchantre** [comment]
No problem.

Maybe it is a simple condition to add in BakerViewController.m near the line 1044?
It seems there's no check if the url contains a referrer or not.

```
if ([[url scheme] isEqualToString:@"file"])
{
    // ****** Handle: file://
    NSLog(@"[BakerView]     Page is a link with scheme file:// --> load internal link");

    anchorFromURL  = [[url fragment] retain];
    NSString *file = [[url relativePath] stringByReplacingPercentEscapesUsingEncoding:NSUTF8StringEncoding];

    int page = [pages indexOfObject:file];
    if (page == NSNotFound)
    {
        // ****** Internal link, but not one of the book pages --> load page anyway
        return YES;
    }

    page = page + 1;
    if (![self changePage:page] && ![webView isEqual:indexViewController.view])
    {
        if (anchorFromURL == nil) {
            return YES;
        }

        [self handleAnchor:YES];
    }
}
```

**Comment by benoitchantre** [comment]
Possible workaround: #944 

Update: doesn't work

**Comment by benoitchantre** [comment]
My latest test with the proper referrer to open a PDF in the modal view replaces the content of the page by the PDF when I click on the link instead to open it in the modal view.

Here's the link I have used for my test in a page named pdf.html
`<a href="pdf/ipad_guide_de_l_utilisateur.pdf?referrer=Baker">Guide de l'utilisateur iPad (lien interne)</a>`

In the log:
`[BakerView] Swiping to page: 5`
`[BakerView] Loading: pdf.html`
`[BakerView] Page is a link with scheme file:// --> load internal link`

Can someone reproduce that or open the pdf in the modal view as expected?

**Comment by benoitchantre** [comment]
Pull request #983
